osgar/lib/depth.py

```diff
@@ -5,97 +5,123 @@
+class DepthParams:
+    def __init__(
+            self,
+            # Focal length.
+            fx=462.1,
+            # Placement of the camera reliative to the center of the robot.
+            camera_xyz=(0.23, 0, (0.19 + 0.06256005)),
+            # Image dimensions.
+            image_size=(640, 360),
+            # Principal point, position of optical axis
+            principal_point=(640 / 2. +  0.5, 360 / 2 + 0.5),
+            max_xy=(16., 16.),
+            # Low-height stuff on the ground does not matter. (In meters.)
+            min_z=0.08,
+            # Stuff above the robot does not matter.
+            max_z=1.5,
+            # We compare pixels this far away from each other vertically.
+            # (In pixels.)
+            vertical_pixel_offset=1,
+            # How close to the vertical does an obstacle need to be?
+            # (In radians.)
+            vertical_diff_limit=np.radians(60),
+            # How much of a slope do we still consider "normal".
+            max_slope=np.radians(-16),
+            # Despeckling parameters.
+            noise_filter_window=(5, 5),
+            noise_filter_threshold=20):
+
+        self.fx = fx
+        self.camera_xyz = np.asarray(camera_xyz)
+        self.camw, self.camh = image_size
+        self.cam_low = self.camh // 2 + 1
+        self.rx, self.ry = principal_point
+        self.max_x, self.max_y = max_xy
+        self.min_z = min_z
+        self.max_z = max_z
+        self.vertical_pixel_offset = vertical_pixel_offset
+        self.vertical_diff_limit = vertical_diff_limit
+        self.max_slope = max_slope
+        self.noise_filter_window = noise_filter_window
+        self.noise_filter_threshold = noise_filter_threshold
+
+        # Pixel coordinates relative to the center of the image, with positive
+        # directions to the left and up.
+        pxs = self.rx - np.repeat(
+                np.arange(self.camw).reshape((1, self.camw)), self.camh, axis=0)
+        pys = self.ry - np.repeat(
+                np.arange(self.camh).reshape((self.camh, 1)), self.camw, axis=1)
+        pzs = np.ones((self.camh, self.camw), dtype=np.float)
+        # For each pixel in the image, a vector representing its corresponding
+        # direction in the scene with a unit forward axis.
+        self.ps = np.dstack([pzs, pxs / fx, pys / fx])
 
 
 # Indices of directions in a matrix with 3D points.
 X, Y, Z = 0, 1, 2
 
-# Pixel coordinates relative to the center of the image, with positive
-# directions to the left and up.
-pxs = RX - np.repeat(np.arange(CAMW).reshape((1, CAMW)), CAMH, axis=0)
-pys = RY - np.repeat(np.arange(CAMH).reshape((CAMH, 1)), CAMW, axis=1)
-pzs = np.ones((CAMH, CAMW), dtype=np.float)
-# For each pixel in the image, a vector representing its corresponding direction
-# in the scene with a unit forward axis.
-ps = np.dstack([pzs, pxs / FX, pys / FX])
-
 
-def depth2danger(depth_mm):
+def depth2danger(depth_mm, params=DepthParams()):
```

Comment:
actually this default param is also not OK - as far as I remember it is created always independently if it is used - I would change it to `None` and then re-initialized it inside, OK?

Reply:
It's the other way around - it created only on import and since the `params` are read-only we are ducking a different problem - if params were to be modified, the modification from one call would persist to other call.

Reply:
Actually, I think it will be best if we don't have default `params` here at all and just fix all callers to supply them directly.